backend/src/packages/articles/article.service.ts

```diff
@@ -1,19 +1,75 @@
 import { ApplicationError } from '~/libs/exceptions/exceptions.js';
+import { safeJSONParse } from '~/libs/helpers/helpers.js';
 import { type IService } from '~/libs/interfaces/service.interface.js';
+import { type OpenAIService } from '~/libs/packages/openai/openai.package.js';
 
+import { GenreEntity } from '../genres/genre.entity.js';
+import { type GenreRepository } from '../genres/genre.repository.js';
 import { ArticleEntity } from './article.entity.js';
 import { type ArticleRepository } from './article.repository.js';
+import { getDetectArticleGenrePrompt } from './libs/helpers/get-detect-article-genre-prompt.helper.js';
 import {
   type ArticleBaseResponseDto,
   type ArticleCreateDto,
   type ArticleUpdateRequestDto,
+  type DetectedArticleGenre,
 } from './libs/types/types.js';
 
 class ArticleService implements IService {
   private articleRepository: ArticleRepository;
+  private openAIService: OpenAIService;
+  private genreRepository: GenreRepository;
 
-  public constructor(articleRepository: ArticleRepository) {
+  public constructor(
+    articleRepository: ArticleRepository,
+    openAIService: OpenAIService,
+    genreRepository: GenreRepository,
+  ) {
     this.articleRepository = articleRepository;
+    this.openAIService = openAIService;
+    this.genreRepository = genreRepository;
+  }
+
+  public async detectArticleGenreFromText(
+    text: string,
+  ): Promise<DetectedArticleGenre | null> {
+    const genresJSON = await this.openAIService.createCompletion({
+      temperature: 0,
+      prompt: getDetectArticleGenrePrompt(text),
+    });
+
+    if (!genresJSON) {
+      return null;
+    }
+
+    const [firstParsedGenre] =
+      safeJSONParse<DetectedArticleGenre[]>(genresJSON) ?? [];
+
+    return firstParsedGenre ?? null;
+  }
+
+  private async getGenreIdForArticle(
+    articleText: string,
+  ): Promise<number | null> {
+    const detectedGenre = await this.detectArticleGenreFromText(articleText);
+
+    if (!detectedGenre) {
+      return null;
+    }
+
+    const existingGenre = await this.genreRepository.findByKey(
+      detectedGenre.key,
+    );
+
+    if (existingGenre) {
+      return existingGenre.toObject().id;
+    } else {
+      const newGenreEntity = await this.genreRepository.create(
+        GenreEntity.initializeNew(detectedGenre),
+      );
+
+      return newGenreEntity.toObject().id;
+    }
```

Comment:
Else block is not necessary as you have return in if block
```suggestion
    if (existingGenre) {
      return existingGenre.toObject().id;
    }
    
    const newGenreEntity = await this.genreRepository.create(
        GenreEntity.initializeNew(detectedGenre),
      );

      return newGenreEntity.toObject().id;
```